### PR TITLE
bpo-30046: [WIP] Cast a Bool object to float type when writing csv.

### PR DESCRIPTION
**Pull request by corona10**
bpo-30046: Cast a Bool object to float type when writing csv.

wip: TODO => Add unittests.

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
`bool` is just one particular case. This patch is not general enough. And it contains a leak.

I think this is a not right way for resolving this issue.

**Comment by corona10** [comment]
@serhiy-storchaka 
Thanks for the review. 
Can I get some good example to handling ref counting?
(Memory leak looks like relate with ref counting which you pointed right?)

File: Modules/_csv.c [hunk 1]
```diff
@@ -1213,6 +1213,19 @@ csv_writerow(WriterObj *self, PyObject *seq)
             append_ok = join_append(self, field, quoted);
             Py_DECREF(field);
         }
+        else if (PyBool_Check(field)) {
+            PyObject *str;
+            double d = PyFloat_AsDouble(field);
+            str = PyFloat_FromDouble(d);
+            str = PyObject_Str(str);
```
Comment (serhiy-storchaka) [inline Modules/_csv.c:1220]: Here is a memory leak. The float object referenced by str is leaked after reassigning str.

File: Modules/_csv.c [hunk 2]
```diff
@@ -1213,6 +1213,19 @@ csv_writerow(WriterObj *self, PyObject *seq)
             append_ok = join_append(self, field, quoted);
             Py_DECREF(field);
         }
+        else if (PyBool_Check(field)) {
+            PyObject *str;
+            double d = PyFloat_AsDouble(field);
+            str = PyFloat_FromDouble(d);
```
Comment (serhiy-storchaka) [inline Modules/_csv.c:1219]: `PyFloat_FromDouble()` can return NULL and raise an exception.

Comment (corona10) [reply Modules/_csv.c:1219]: @serhiy-storchaka 
Thank you for your kind review.
So in this case, This code should be go this way?

````
            PyObject *str;
            double d = PyFloat_AsDouble(field);
            str = PyFloat_FromDouble(d);
            if (str == NULL) {
                Py_DECREF(iter);
                return NULL;
            }
            str = PyObject_Str(str);
            Py_DECREF(str);
            Py_DECREF(field);
            if (str == NULL) {
                Py_DECREF(iter);
                return NULL;
            }
            append_ok = join_append(self, str, quoted);
            Py_DECREF(str);
````

Comment (corona10) [reply Modules/_csv.c:1219]: If this is correct, I will not make same mistake in a future :-)

Comment (serhiy-storchaka) [reply Modules/_csv.c:1219]:
```
            str = PyObject_Str(str);
            Py_DECREF(str);
```
You have made even worse mistake. Not just the float object is leaked, but the str object is decrefed twice. Use different variables for saving results of `PyFloat_FromDouble()` and `PyObject_Str()`.

Comment (corona10) [reply Modules/_csv.c:1219]: @serhiy-storchaka 
Thanks! I think that I should read related articles.
Thank you for spend times for me.

**Pull request closed without merging**